server/src/main/java/org/elasticsearch/index/query/support/AutoPrefilteringUtils.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.query.support;
+
+import org.elasticsearch.common.lucene.search.Queries;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.BoostingQueryBuilder;
+import org.elasticsearch.index.query.ConstantScoreQueryBuilder;
+import org.elasticsearch.index.query.DisMaxQueryBuilder;
+import org.elasticsearch.index.query.InterceptedQueryBuilderWrapper;
+import org.elasticsearch.index.query.NestedQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.functionscore.FunctionScoreQueryBuilder;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public final class AutoPrefilteringUtils {
+
+    private AutoPrefilteringUtils() {}
+
+    /**
+     * Prunes any query branch on queries whose type is included in the prunedTypes set.
+     * Note that queries may not preserve their scoring parameters as this assumes the returned
+     * query is going to be used for filtering only.
+     *
+     * @param query the root query to start pruning from
+     * @param prunedTypes the types of queries to prune
+     * @return an {@link Optional} containing the pruned query, or {@link Optional#empty()} if the query was pruned to nothing.
+     */
+    public static Optional<QueryBuilder> pruneQuery(QueryBuilder query, Set<Class<? extends QueryBuilder>> prunedTypes) {
+        if (prunedTypes.contains(query.getClass())) {
+            // Matched a pruned query type, prune away!
+            return Optional.empty();
+        }
+
+        if (query instanceof BoolQueryBuilder boolQuery) {
```

Comment:
Nit - Consider using [pattern matching for switch ](https://docs.oracle.com/en/java/javase/17/language/pattern-matching-switch.html)